File: app/src/main/java/arcus/app/launch/InvitationFragment.java
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.launch;

import android.app.Activity;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import arcus.cornea.CorneaClientFactory;
import arcus.cornea.SessionController;
import arcus.cornea.utils.Listeners;
import com.iris.client.capability.Person;
import com.iris.client.event.Listener;
import com.iris.client.exception.ErrorResponseException;
import com.iris.client.model.PersonModel;
import com.iris.client.service.InvitationService;
import arcus.app.R;
import arcus.app.account.registration.controller.AccountCreationSequenceController;
import arcus.app.account.registration.model.AccountTypeSequence;
import arcus.app.account.settings.SettingsUpdatePin;
import arcus.app.account.settings.SideNavSettingsFragment;
import arcus.app.activities.LaunchActivity;
import arcus.app.common.backstack.BackstackManager;
import arcus.app.common.backstack.TransitionEffect;
import arcus.app.common.error.ErrorManager;
import arcus.app.common.fragments.BaseFragment;
import arcus.app.common.image.ImageManager;
import arcus.app.common.image.Wallpaper;
import arcus.app.common.popups.AlertPopup;
import arcus.app.common.popups.InfoButtonPopup;
import arcus.app.common.utils.PreferenceUtils;
import arcus.app.common.validation.EmailValidator;
import arcus.app.common.validation.NotEmptyValidator;
import arcus.app.common.view.Version1Button;
import arcus.app.common.view.Version1ButtonColor;
import arcus.app.common.view.Version1EditText;
import arcus.app.subsystems.alarm.AlertFloatingFragment;
import arcus.app.subsystems.people.model.DeviceContact;

import java.util.Map;


public class InvitationFragment extends BaseFragment {
    private static String EMAIL_ADDRESS="EMAIL_ADDRESS";
    private static String INVITATION_CODE="INVITATION_CODE";
    public static String INVITATION_FIRST_NAME="INVITATION_FIRST_NAME";
    public static String INVITATION_LAST_NAME="INVITATION_LAST_NAME";
    public static String IS_SETTINGS="IS_SETTINGS";

    Version1EditText etEmailAddress;
    Version1EditText etInvitation;
    Version1Button next;
    AlertFloatingFragment alertpopup;

    private boolean isSettingsVariant;
    final Listener<Throwable> genericErrorListener = Listeners.runOnUiThread(new Listener<Throwable>() {
        @Override public void onEvent(Throwable throwable) {
            hideProgressBarAndEnable(next);
            if (throwable instanceof ErrorResponseException) {
                parseErrorException((ErrorResponseException) throwable);
            }
            else {
                onError(throwable);
            }
        }
    });

    @NonNull
    public static InvitationFragment newInstance(String emailAddress, String invitationCode, String firstName, String lastName){
        InvitationFragment fragment = new InvitationFragment();
        final Bundle arguments = new Bundle();
        arguments.putString(EMAIL_ADDRESS, emailAddress);
        arguments.putString(INVITATION_CODE, invitationCode);
        arguments.putString(INVITATION_FIRST_NAME, firstName);
        arguments.putString(INVITATION_LAST_NAME, lastName);
        fragment.setArguments(arguments);
        return fragment;
    }

    @NonNull public static InvitationFragment newInstanceFromSettings() {
        InvitationFragment fragment = new InvitationFragment();
        final Bundle arguments = new Bundle(1);
        arguments.putBoolean(IS_SETTINGS, true);
        fragment.setArguments(arguments);
        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        isSettingsVariant = getArguments().getBoolean(IS_SETTINGS, false);
        View view = super.onCreateView(inflater, container, savedInstanceState);
        String emailAddress = getArguments().getString(EMAIL_ADDRESS);
        String invitationCode = getArguments().getString(INVITATION_CODE);

        etEmailAddress = (Version1EditText) view.findViewById(R.id.etEmail);
        etInvitation = (Version1EditText) view.findViewById(R.id.etInvitation);

        if(emailAddress != null) {
            etEmailAddress.setText(emailAddress);
        }
        if(invitationCode != null) {
            etInvitation.setText(invitationCode);
        }

        if (!isSettingsVariant) {
            ImageManager.with(getActivity()).setWallpaper(Wallpaper.ofDefaultWallpaper().lightend());
        }

        setTitleOfPage();
        next = (Version1Button) view.findViewById(R.id.btnNext);
        next.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                boolean emailValid = new EmailValidator(etEmailAddress).isValid();
                boolean inviteCodeCheck = new NotEmptyValidator(getActivity(), etInvitation).isValid();
                if (!emailValid || !inviteCodeCheck) {
                    return;
                }



                    showProgressBarAndDisable(next);
                if(TextUtils.isEmpty(CorneaClientFactory.getClient().getConnectionURL())) {
                    CorneaClientFactory.getClient().setConnectionURL(PreferenceUtils.getPlatformUrl());
                }
                    CorneaClientFactory.getService(InvitationService.class)
                            .getInvitation(etInvitation.getText().toString(), etEmailAddress.getText().toString())
                            .onSuccess(Listeners.runOnUiThread(new Listener<InvitationService.GetInvitationResponse>() {
                                @Override
                                public void onEvent(InvitationService.GetInvitationResponse getInvitationResponse) {
                                    Map<String, Object> invite = getInvitationResponse.getInvitation();
                                    if (invite == null) {
                                        onError(new RuntimeException("Invite was null, but no error from server."));
                                        return;
                                    }

                                    String first = (String) invite.get("invitorFirstName");
                                    String last = (String) invite.get("invitorLastName");
                                    String personName = String.format("%s %s", first == null ? "" : first, last == null ? "" : last);

                                    String placeName = (String) invite.get("placeName");
                                    placeName = placeName == null ? "" : placeName;

                                    String inviteeEmail = (String) invite.get("inviteeEmail");

                                    String placeID = String.valueOf(invite.get("placeId"));
                                    if (isSettingsVariant) {
                                        showAccept(placeName, personName, placeID, inviteeEmail);
                                    }
                                    else {
                                        String inviteeFirst = (String) invite.get("inviteeFirstName");
                                        String inviteeLast = (String) invite.get("inviteeLastName");
                                        DeviceContact contact = new DeviceContact();
                                        contact.addEmailAddress(etEmailAddress.getText().toString(), getResources().getString(R.string.type_home));
                                        contact.setFirstName(inviteeFirst);
                                        contact.setLastName(inviteeLast);
                                        contact.setPlaceID(placeID);
                                        contact.setValidationCode(etInvitation.getText().toString());
                                        contact.setInvitationEmail(inviteeEmail);
                                        contact.setInvitedPlaceName(placeName);
                                        contact.setInvitorFirstName(first);
                                        contact.setInvitorLastName(last);
                                        BackstackManager.getInstance().navigateToFragment(InvitationSuccessFragment.newInstance(contact), true);
                                    }
                                }
                            }))
                            .onFailure(Listeners.runOnUiThread(new Listener<Throwable>() {
                                @Override
                                public void onEvent(Throwable throwable) {
                                    showInvalid();
                                }
                            }));
            }
        });
        return view;
    }

    protected void showAccept(String toPlace, String fromPerson, final String forPlaceID, final String inviteeEmail) {
        hideProgressBarAndEnable(next);


        InfoButtonPopup popup = InfoButtonPopup.newInstance(
                getString(R.string.accept_invitation_title).toUpperCase(),
                getString(R.string.accept_invitation_desc, toPlace, fromPerson),
                getString(R.string.accept),
                getString(R.string.decline),
                Version1ButtonColor.MAGENTA,
                Version1ButtonColor.BLACK
        );

        // popup callback
        popup.setCallback(new InfoButtonPopup.Callback() {
            @Override public void confirmationValue(boolean topButton) {

                final PersonModel personModel = SessionController.instance().getPerson();
                if (personModel == null) {
                    onError(new RuntimeException("Lost our logged in person. Cannot accept/decline invite."));
                    return;
                }

                showProgressBarAndDisable(next);
                String code = etInvitation.getText().toString();
                String email = etEmailAddress.getText().toString();

                if (topButton) { // Accept
                    personModel.acceptInvitation(code, email)
                               .onFailure(genericErrorListener)
                               .onSuccess(Listeners.runOnUiThread(new Listener<Person.AcceptInvitationResponse>() {
                                   @Override
                                   public void onEvent(Person.AcceptInvitationResponse acceptInvitationResponse) {
                                       DeviceContact contact = new DeviceContact();
                                       contact.addEmailAddress(etEmailAddress.getText().toString(),
                                               getResources().getString(R.string.type_home));
                                       contact.setFirstName(personModel.getFirstName());
                                       contact.setLastName(personModel.getLastName());
                                       contact.setPlaceID(forPlaceID);
                                       contact.setValidationCode(etInvitation.getText().toString());
                                       contact.setInvitationEmail(inviteeEmail);

                                       new AccountCreationSequenceController(
                                               AccountTypeSequence.CURRENT_USER_INVITE_ACCEPT,
                                               contact).startSequence(getActivity(), null, SettingsUpdatePin.class);
                                   }
                               }));
                } else { // Decline
                    personModel.rejectInvitation(code, email, null) // No "Reason" for now.
                               .onFailure(genericErrorListener)
                               .onSuccess(Listeners.runOnUiThread(new Listener<Person.RejectInvitationResponse>() {
                                   @Override public void onEvent(Person.RejectInvitationResponse rejectInvitationResponse) {
                                       hideProgressBar();BackstackManager.withAnimation(TransitionEffect.FADE).navigateBackToFragment(
                                               SideNavSettingsFragment.newInstance());
                                   }
                               }));
                }
            }
        });
        BackstackManager.getInstance().navigateToFloatingFragment(popup, popup.getClass().getSimpleName(), true);
    }

    protected void onError(Throwable throwable) {
        ErrorManager.in(getActivity()).showGenericBecauseOf(throwable);
    }

    protected void parseErrorException(ErrorResponseException ex) {
        String NOT_FOUND = "request.destination.notfound";
        String prefix = "place";
        String code = String.valueOf(ex.getCode()).toLowerCase();
        String reason = String.valueOf(ex.getErrorMessage()).toLowerCase().trim();
        if (code.equals(NOT_FOUND) && reason.startsWith(prefix)) {
            AlertPopup alert = AlertPopup.newInstance(
                  getString(R.string.invite_not_valid_title),
                  getString(R.string.invite_not_valid_desc),
                  null, new AlertPopup.AlertButtonCallback() {
                      @Override public boolean topAlertButtonClicked() { return false; }
                      @Override public boolean bottomAlertButtonClicked() { return false; }
                      @Override public boolean errorButtonClicked() { return false; }
                      @Override public void close() {
                          BackstackManager.getInstance().navigateBack();
                      }
                  }
            );
            BackstackManager.getInstance().navigateToFloatingFragment(alert, alert.getClass().getCanonicalName(), true);
        }
        else {
            onError(ex);
        }
    }

    protected void showInvalid() {
        hideProgressBarAndEnable(next);
        String notValidText = isSettingsVariant ? getString(R.string.code_not_valid_desc) : getString(R.string.code_not_valid_short_desc);
        alertpopup = AlertFloatingFragment.newInstance(
              getString(R.string.code_not_valid_title),
              notValidText,
              null, null, null
        );
        BackstackManager.getInstance().navigateToFloatingFragment(alertpopup, alertpopup.getClass().getSimpleName(), true);
    }

    @Override public void onPause() {
        super.onPause();
        hideProgressBarAndEnable(next);
    }

    protected void setTitleOfPage() {
        Activity activity = getActivity();
        if (activity == null) {
            return;
        }

        activity.setTitle(getTitle());
    }

    @Nullable @Override public String getTitle() {
        return isSettingsVariant ? getString(R.string.invitation_code) : getResources().getString(R.string.invitation_title);
    }

    @Override
    public Integer getLayoutId() {
        return isSettingsVariant ? R.layout.invite_settings_fragment : R.layout.fragment_invitation;
    }

    public void handleBackPress() {
        if (popupVisible()) {
            BackstackManager.getInstance().navigateBack();
        } else {
            LaunchActivity.startLoginScreen(getActivity());
        }
    }

    public boolean popupVisible() {
        if (alertpopup != null && alertpopup.isVisible()) {
            return true;
        }
        return false;
    }
}

File: app/src/main/java/arcus/app/launch/InvitationFragment.kt
/*
 *  Copyright 2019 Arcus Project.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package arcus.app.launch

import android.os.Bundle
import android.text.TextUtils
import android.view.View
import android.widget.Button
import android.widget.EditText
import androidx.core.view.isGone
import androidx.core.view.isVisible
import arcus.app.R
import arcus.app.account.registration.controller.AccountCreationSequenceController
import arcus.app.account.registration.model.AccountTypeSequence
import arcus.app.account.settings.pin.SettingsUpdatePin
import arcus.app.account.settings.list.SideNavSettingsFragment
import arcus.app.activities.LaunchActivity
import arcus.app.common.backstack.BackstackManager
import arcus.app.common.backstack.TransitionEffect
import arcus.app.common.error.ErrorManager.`in` as errorIn
import arcus.app.common.fragments.GenericInformationPopup
import arcus.app.common.fragments.NoViewModelFragment
import arcus.app.common.utils.PreferenceUtils
import arcus.app.common.validation.EmailValidator
import arcus.app.common.validation.NotEmptyValidator
import arcus.app.pairing.hub.popups.Generic1ButtonErrorPopup
import arcus.app.subsystems.people.model.DeviceContact
import arcus.cornea.CorneaClientFactory
import arcus.cornea.SessionController
import arcus.cornea.utils.Listeners
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textfield.TextInputLayout
import com.iris.client.exception.ErrorResponseException
import com.iris.client.service.InvitationService
import com.iris.client.service.InvitationService.GetInvitationResponse
import java.util.Locale

// TODO: This class was just converted to Kotlin and still needs significant work to update it.
class InvitationFragment : NoViewModelFragment() {
    private lateinit var etEmailAddress: EditText
    private lateinit var etInvitation: EditText
    private lateinit var emailEntryContainer: TextInputLayout
    private lateinit var invitationCodeContainer: TextInputLayout
    private lateinit var next: Button
    private var emailAddress: String? = null
    private var invitationCode: String? = null
    private var isSettingsVariant = false
    private val genericErrorListener = Listeners.runOnUiThread<Throwable> { throwable ->
        progressContainer.isGone = true
        next.isEnabled = true
        if (throwable is ErrorResponseException) {
            parseErrorException(throwable)
        } else {
            onError(throwable)
        }
    }

    override val title: String
        get() = if (isSettingsVariant) {
            getString(R.string.invitation_code)
        } else {
            resources.getString(R.string.invitation_title)
        }
    override val layoutId: Int
        get() = if (isSettingsVariant) {
            R.layout.invite_settings_fragment
        } else {
            R.layout.fragment_invitation
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        arguments?.let { args ->
            emailAddress = args.getString(EMAIL_ADDRESS)
            invitationCode = args.getString(INVITATION_CODE)
            isSettingsVariant = args.getBoolean(IS_SETTINGS, false)
        }
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        etEmailAddress = view.findViewById(R.id.emailEntry)
        etInvitation = view.findViewById(R.id.invitationCodeEntry)
        emailEntryContainer = view.findViewById(R.id.emailContainer)
        invitationCodeContainer = view.findViewById(R.id.invitationCodeContainer)

        etEmailAddress.setText(emailAddress)
        etInvitation.setText(invitationCode)

        next = view.findViewById(R.id.btnNext)
        next.setOnClickListener(View.OnClickListener {
            val emailValid = EmailValidator(emailEntryContainer, etEmailAddress).isValid
            val inviteCodeCheck = NotEmptyValidator(
                invitationCodeContainer,
                etInvitation,
                getString(R.string.people_err_please_enter)
            ).isValid
            if (!emailValid || !inviteCodeCheck) {
                return@OnClickListener
            }
            progressContainer.isVisible = true
            next.isEnabled = false
            if (TextUtils.isEmpty(CorneaClientFactory.getClient().connectionURL)) {
                CorneaClientFactory.getClient().connectionURL = PreferenceUtils.getPlatformUrl()
            }
            CorneaClientFactory
                .getService(InvitationService::class.java)
                .getInvitation(etInvitation.text.toString(), etEmailAddress.text.toString())
                .onSuccess(
                    Listeners.runOnUiThread { getInvitationResponse: GetInvitationResponse ->
                        val invite =
                            getInvitationResponse.invitation
                        if (invite == null) {
                            onError(RuntimeException("Invite was null, but no error from server."))
                            return@runOnUiThread
                        }
                        val first = invite["invitorFirstName"] as String?
                        val last = invite["invitorLastName"] as String?
                        val personName = String.format(
                            "%s %s",
                            first ?: "",
                            last ?: ""
                        )
                        var placeName = invite["placeName"] as String?
                        placeName = placeName ?: ""
                        val inviteeEmail = invite["inviteeEmail"] as String?
                        val placeID = invite["placeId"].toString()
                        if (isSettingsVariant) {
                            showAccept(placeName, personName, placeID, inviteeEmail)
                        } else {
                            val inviteeFirst = invite["inviteeFirstName"] as String?
                            val inviteeLast = invite["inviteeLastName"] as String?
                            val contact = DeviceContact()
                            contact.addEmailAddress(
                                etEmailAddress.text.toString(),
                                resources.getString(R.string.type_home)
                            )
                            contact.firstName = inviteeFirst
                            contact.lastName = inviteeLast
                            contact.placeID = placeID
                            contact.validationCode = etInvitation.text.toString()
                            contact.invitationEmail = inviteeEmail
                            contact.invitedPlaceName = placeName
                            contact.invitorFirstName = first
                            contact.invitorLastName = last
                            BackstackManager
                                .getInstance()
                                .navigateToFragment(InvitationSuccessFragment.newInstance(contact), true)
                        }
                    })
                .onFailure(Listeners.runOnUiThread { showInvalid() })
        })
    }

    private fun showAccept(
        toPlace: String,
        fromPerson: String,
        forPlaceID: String,
        inviteeEmail: String?
    ) {
        progressContainer.isGone = true
        next.isEnabled = true
        val code = etInvitation.text.toString()
        val email = etEmailAddress.text.toString()
        GenericInformationPopup
            .newInstance(
                getString(R.string.accept_invitation_title),
                getString(R.string.accept_invitation_desc, toPlace, fromPerson),
                getString(R.string.accept),
                getString(R.string.decline)
            )
            .setTopButtonListener {
                val personModel = SessionController.instance().person
                if (personModel == null) {
                    onError(RuntimeException("Lost our logged in person. Cannot accept/decline invite."))
                    return@setTopButtonListener
                }
                progressContainer.isVisible = true
                next.isEnabled = false
                personModel
                    .acceptInvitation(code, email)
                    .onFailure(genericErrorListener)
                    .onSuccess(
                        Listeners.runOnUiThread {
                            val contact = DeviceContact()
                            contact.addEmailAddress(
                                etEmailAddress.text.toString(),
                                resources.getString(R.string.type_home)
                            )
                            contact.firstName = personModel.firstName
                            contact.lastName = personModel.lastName
                            contact.placeID = forPlaceID
                            contact.validationCode = etInvitation.text.toString()
                            contact.invitationEmail = inviteeEmail
                            AccountCreationSequenceController(
                                AccountTypeSequence.CURRENT_USER_INVITE_ACCEPT,
                                contact
                            ).startSequence(activity, null, SettingsUpdatePin::class.java)
                        })
            }
            .setBottomButtonListener {
                val personModel = SessionController.instance().person
                if (personModel == null) {
                    onError(RuntimeException("Lost our logged in person. Cannot accept/decline invite."))
                    return@setBottomButtonListener
                }
                progressContainer.isVisible = true
                next.isEnabled = false
                personModel
                    .rejectInvitation(code, email, null) // No "Reason" for now.
                    .onFailure(genericErrorListener)
                    .onSuccess(
                        Listeners.runOnUiThread {
                            // TODO: Should this really go back to here?...
                            progressContainer.isGone = true
                            BackstackManager
                                .withAnimation(TransitionEffect.FADE)
                                .navigateBackToFragment(SideNavSettingsFragment.newInstance())
                        })
            }
            .show(fragmentManager)
    }

    private fun onError(throwable: Throwable?) {
        errorIn(activity).showGenericBecauseOf(throwable)
    }

    private fun parseErrorException(ex: ErrorResponseException) {
        val NOT_FOUND = "request.destination.notfound"
        val prefix = "place"
        val code = ex.code.toString().toLowerCase(Locale.getDefault())
        val reason = ex.errorMessage.toString().toLowerCase(Locale.getDefault()).trim { it <= ' ' }
        if (code == NOT_FOUND && reason.startsWith(prefix)) {
            MaterialAlertDialogBuilder(requireContext())
                .setTitle(R.string.invite_not_valid_title)
                .setMessage(R.string.invite_not_valid_desc)
                .setPositiveButton(android.R.string.ok, null)
                .create()
                .show()
        } else {
            onError(ex)
        }
    }

    private fun showInvalid() {
        progressContainer.isGone = true
        next.isEnabled = true
        val notValidText = if (isSettingsVariant) {
            getString(R.string.code_not_valid_desc)
        } else {
            getString(R.string.code_not_valid_short_desc)
        }
        Generic1ButtonErrorPopup
            .newInstance(
                getString(R.string.code_not_valid_title),
                notValidText,
                "",
                "",
                getString(android.R.string.ok)
            )
            .show(fragmentManager)
    }

    override fun onPause() {
        super.onPause()
        progressContainer.isGone = true
        next.isEnabled = true
    }

    fun handleBackPress() { // Invoked from BaseActivity only it seems.
        LaunchActivity.startLoginScreen(activity)
    }

    companion object {
        private const val EMAIL_ADDRESS = "EMAIL_ADDRESS"
        private const val INVITATION_CODE = "INVITATION_CODE"
        const val IS_SETTINGS = "IS_SETTINGS"

        @JvmStatic
        fun newInstance(
            emailAddress: String?,
            invitationCode: String?,
            firstName: String?,
            lastName: String?
        ): InvitationFragment = InvitationFragment().apply {
            arguments = with(Bundle(2)) {
                putString(EMAIL_ADDRESS, emailAddress)
                putString(INVITATION_CODE, invitationCode)
                this
            }
        }

        @JvmStatic
        fun newInstanceFromSettings(): InvitationFragment = InvitationFragment().apply {
            arguments = with(Bundle(1)) {
                putBoolean(IS_SETTINGS, true)
                this
            }
        }
    }
}
